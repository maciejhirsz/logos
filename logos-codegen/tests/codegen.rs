use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
