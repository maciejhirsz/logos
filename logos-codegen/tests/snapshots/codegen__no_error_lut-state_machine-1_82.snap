---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = [u8];
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use ::logos::internal::{
            LexerInternal, CallbackRetVal, CallbackResult, SkipRetVal, SkipResult,
        };
        use std::result::Result as _Result;
        use std::option::Option as _Option;
        use ::logos::Logos;
        type _Lexer<'s> = ::logos::Lexer<'s, Token>;
        #[derive(Clone, Copy)]
        enum LogosState {
            state10240,
            state1024_ctx2,
            state10752,
            state1536_ctx2,
            state2048_ctx1,
            state2560_ctx2,
            state3072_ctx2,
            state3584_ctx0,
            state4096_ctx1,
            state4608,
            state5120,
            state5632,
            state6144,
            state6656,
            state7168,
            state7680_ctx2,
            state8192_ctx2,
            state8704,
            state9216,
            state9728,
        }
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <Token as Logos<'s>>::Error {
            <Token as Logos<'s>>::Error::default()
        }
        let mut state = LogosState::state4608;
        let mut offset = lex.offset();
        loop {
            match state {
                LogosState::state1024_ctx2 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state1536_ctx2 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state7680_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state2048_ctx1 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::AnyUnicode));
                        }
                    }
                }
                LogosState::state2560_ctx2 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state8192_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state3072_ctx2 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        Some(0u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state2048_ctx1;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state2048_ctx1;
                                continue;
                            }
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state3584_ctx0 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::Newline));
                        }
                    }
                }
                LogosState::state4096_ctx1 => {
                    lex.end(offset - 1);
                    match lex.read::<u8>(offset) {
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::AnyUnicode));
                        }
                    }
                }
                LogosState::state4608 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        const TABLE: [_Option<LogosState>; 256] = [
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state8704),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state7168),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state9728),
                            Some(LogosState::state10240),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state6656),
                            Some(LogosState::state10752),
                            Some(LogosState::state10752),
                            Some(LogosState::state5120),
                            Some(LogosState::state5632),
                            Some(LogosState::state5632),
                            Some(LogosState::state5632),
                            Some(LogosState::state6144),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                            Some(LogosState::state9216),
                        ];
                        let next_state = TABLE[byte as usize];
                        if let Some(next_state) = next_state {
                            offset += 1;
                            state = next_state;
                            continue;
                        }
                    } else {
                        if lex.offset() == offset {
                            return None;
                        }
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
                LogosState::state5120 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=143u8) | Some(192u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(144u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state1536_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state5632 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=127u8) | Some(192u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state1536_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state6144 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=127u8) | Some(144u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(128u8..=143u8) => {
                            offset += 1;
                            state = LogosState::state1536_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state6656 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=127u8) | Some(160u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(128u8..=159u8) => {
                            offset += 1;
                            state = LogosState::state2560_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state7168 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state4096_ctx1;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state4096_ctx1;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state7680_ctx2 => {
                    match lex.read::<u8>(offset) {
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state8192_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {}
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state8192_ctx2 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state2048_ctx1;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state2048_ctx1;
                                continue;
                            }
                            return Some(Ok(Token::Any));
                        }
                    }
                }
                LogosState::state8704 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state3584_ctx0;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state3584_ctx0;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state9216 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state9728 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=127u8) | Some(192u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state3072_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state10240 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=159u8) | Some(192u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(160u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state2560_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
                LogosState::state10752 => {
                    match lex.read::<u8>(offset) {
                        Some(0u8..=127u8) | Some(192u8..=255u8) => {
                            offset += 1;
                            state = LogosState::state1024_ctx2;
                            continue;
                        }
                        Some(128u8..=191u8) => {
                            offset += 1;
                            state = LogosState::state2560_ctx2;
                            continue;
                        }
                        other => {
                            if other.is_none() {
                                offset += 1;
                                state = LogosState::state1024_ctx2;
                                continue;
                            }
                            lex.end_to_boundary(
                                offset + if other.is_some() { 1 } else { 0 },
                            );
                            return Some(Err(_make_error(lex)));
                        }
                    }
                }
            }
        }
    }
}
